lib/client.js

```diff
@@ -2096,6 +2162,125 @@ MatrixClient.prototype.searchMessageText = function(opts, callback) {
 };
 
 /**
+ * Perform a server-side search for room events.
+ *
+ * The returned promise resolves to an object containing the fields:
+ *
+ *  * {number}  count:       estimate of the number of results
+ *  * {string}  next_batch:  token for back-pagination; if undefined, there are
+ *                           no more results
+ *  * {Array}   highlights:  a list of words to highlight from the stemming
+ *                           algorithm
+ *  * {Array}   results:     a list of results
+ *
+ * Each entry in the results list is a {module:models/search-result.SearchResult}.
+ *
+ * @param {Object} opts
+ * @param {string} opts.term     the term to search for
+ * @param {Object} opts.filter   a JSON filter object to pass in the request
+ * @return {module:client.Promise} Resolves: result object
+ * @return {module:http-api.MatrixError} Rejects: with an error response.
+ */
+MatrixClient.prototype.searchRoomEvents = function(opts) {
+    // TODO: support groups
+
+    var body = {
+        search_categories: {
+            room_events: {
+                search_term: opts.term,
+                filter: opts.filter,
+                order_by: "recent",
+                event_context: {
+                    before_limit: 1,
+                    after_limit: 1,
+                    include_profile: true,
+                }
+            }
+        }
+    };
+
+    var searchResults = {
+        _query: body,
+        results: [],
+        highlights: [],
+    };
+
+    return this.search({body: body}).then(
+        this._processRoomEventsSearch.bind(this, searchResults)
+    );
+};
+
+/**
+ * Take a result from an earlier searchRoomEvents call, and backfill results.
+ *
+ * @param  {module:models/search-results.SearchResults} searchResults   the
+ *    results object to be updated
+ * @return {module:client.Promise} Resolves: updated result object
+ * @return {module:http-api.MatrixError} Rejects: with an error response.
+ */
+MatrixClient.prototype.backPaginateRoomEventsSearch = function(searchResults) {
+    // TODO: we should implement a backoff (as per scrollback()) to deal more
+    // nicely with HTTP errors.
+
+    if (!searchResults.next_batch) {
+        throw new Error("Cannot backpaginate event search any further");
+    }
+
+    if (searchResults.pendingRequest) {
+        // already a request in progress - return the existing promise
+        return searchResults.pendingRequest;
+    }
+
+    var searchOpts = {
+        body: searchResults._query,
+        next_batch: searchResults.next_batch,
+    };
+
+    var promise = this.search(searchOpts).then(
+        this._processRoomEventsSearch.bind(this, searchResults)
+    ).finally(function() {
+        searchResults.pendingRequest = null;
+    });
```

Comment:
Oh but this one is interesting because you're then shoving the promise away, hmm :/ you'll need to do `promise.done()` stand-alone then.